mlir/lib/Quantum/Transforms/wrapper_for_wrapper.cpp

```diff
@@ -0,0 +1,220 @@
+// Copyright 2022-2023 Xanadu Quantum Technologies Inc.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "mlir/Conversion/LLVMCommon/TypeConverter.h"
+#include "mlir/Dialect/LLVMIR/LLVMDialect.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "llvm/Support/FormatVariadic.h"
+
+#include "Quantum/Transforms/Passes.h"
+#include "Quantum/Transforms/Patterns.h"
+
+using namespace mlir;
+using namespace catalyst::quantum;
+
+namespace catalyst {
+namespace quantum {
```

Comment:
~~This could also live top-level in the `catalyst` namespace since it's neither related to the quantum or gradient dialect.~~

Actually, the pass creation function already lives in the `catalyst` namespace, so all this other functionality would normally live in an anonymous namespace (C++ style), rather than marking them all static (C style).

```suggestion
namespace {
```

Reply:
https://github.com/PennyLaneAI/catalyst/pull/106/commits/723b2f63e6f83c59410a5969d4a91955ea5f3b06

Reply:
I think everything but the pass creation function can go in the anonymous namespace no?

Reply:
This is as minimal as it gets. The class also has to be part of the namespace. https://github.com/PennyLaneAI/catalyst/pull/106/commits/60707040b2efb1e33fe92d9d0b32010e0a1532d5

Reply:
https://github.com/PennyLaneAI/catalyst/pull/106/commits/f50e0a3c4197cbe00dd3ef07da622b77a7cf79b1

Reply:
Huh how come the class has to be part of the namespace? Shouldn't be the case, just like [here](https://github.com/PennyLaneAI/catalyst/blob/main/mlir/lib/Gradient/Transforms/ConversionPatterns.cpp) for example.

One more thing, the `static` keyword can be removed from all the functions :) (unless they are actually static methods)

Reply:
There's an error message that points to the following possible solution, but I think we should just keep it as is.

```cpp
/// C++ RTTI is a notoriously difficult topic; given the nature of shared
/// libraries many different approaches fundamentally break down in either the
/// area of support (i.e. only certain types of classes are supported), or in
/// terms of performance (e.g. by using string comparison). This class intends
/// to strike a balance between performance and the setup required to enable its
/// use.
///
/// Assume we are adding support for some class Foo, below are the set of ways
/// in which a given c++ type may be supported:
///
///  * Explicitly via `MLIR_DECLARE_EXPLICIT_TYPE_ID` and
///    `MLIR_DEFINE_EXPLICIT_TYPE_ID`
///
///    - This method explicitly defines the type ID for a given type using the
///      given macros. These should be placed at the top-level of the file (i.e.
///      not within any namespace or class). This is the most effective and
///      efficient method, but requires explicit annotations for each type.
///
///      Example:
///
///        // Foo.h
///        MLIR_DECLARE_EXPLICIT_TYPE_ID(Foo);
///
///        // Foo.cpp
///        MLIR_DEFINE_EXPLICIT_TYPE_ID(Foo);
```

Reply:
Removed `static` here https://github.com/PennyLaneAI/catalyst/pull/106/commits/d61661fb04ad0865b51a73465786ca70c4e801fc

Reply:
Good to know!